File: api/_engine.ts
/* Rating engine: replays the game log chronologically through openskill.js.
 *
 * TypeScript port of backend/engine.py. Model: Bradley-Terry (full pairing) with
 * margin-of-victory enabled (openskill's `bradleyTerryFull` + `score` + `margin`).
 * Scores are normalized to an 11-point scale before rating so that a 21-15 game and
 * an 11-8 game carry comparable margin information.
 *
 * Note vs. the old Python engine: openskill.js applies margin as an external,
 * monotonic multiplier on the rating delta (log1p(diff - margin)), whereas
 * openskill.py folded it into the win-probability term (log1p(diff / margin)).
 * Both make blowouts move ratings more; the JS form is monotonic, so it does NOT
 * exhibit the old "inverse blowout" quirk among upsets. Behavior is fundamentally
 * the same — conservative mu-3sigma leaderboard, margin-aware updates.
 */
import { predictWin, rate, rating } from "openskill";
import { bradleyTerryFull } from "openskill/models";
import type {
  Game,
  HeadToHead,
  LeagueStats,
  Player,
  PlayerStats,
  ProcessedGame,
  RatingSnapshot,
} from "../src/types";
import type { ApiState } from "../src/engine";

export const PROVISIONAL_THRESHOLD = 5;
export const MARGIN = Number(process.env.MARGIN ?? "2.0");

type OSRating = { mu: number; sigma: number };
const RATE_OPTS = { model: bradleyTerryFull, margin: MARGIN } as const;

const snap = (r: OSRating): RatingSnapshot => ({
  mu: r.mu,
  sigma: r.sigma,
  exposed: r.mu - 3 * r.sigma,
});

/** Scale scores to an 11-point-equivalent so margins are comparable across game lengths. */
const normScores = (aScore: number, bScore: number, target: number): [number, number] => {
  const k = 11 / target;
  return [aScore * k, bScore * k];
};

const hist = (s: RatingSnapshot) => ({ exposed: s.exposed, mu: s.mu });

const avg = (xs: number[]): number | null =>
  xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : null;

const winProb = (a: OSRating, b: OSRating): number => predictWin([[a], [b]])[0];

export function compute(players: Player[], games: Game[]): ApiState {
  const ratings = new Map<string, OSRating>(players.map((p) => [p.id, rating()]));
  const history = new Map<string, PlayerStats["ratingHistory"]>(
    players.map((p) => [
      p.id,
      [{ gameIndex: 0, playedAt: p.createdAt, ...hist(snap(ratings.get(p.id)!)) }],
    ]),
  );

  const processed: ProcessedGame[] = [];
  const ordered = [...games].sort((x, y) => x.playedAt - y.playedAt);
  ordered.forEach((g, idx) => {
    const ra = ratings.get(g.aId);
    const rb = ratings.get(g.bId);
    if (!ra || !rb) return;

    const aWon = g.aScore > g.bScore;
    const winnerId = aWon ? g.aId : g.bId;
    const loserId = aWon ? g.bId : g.aId;
    const winnerR = aWon ? ra : rb;
    const loserR = aWon ? rb : ra;

    const winnerPreProb = winProb(winnerR, loserR);
    const aBefore = snap(ra);
    const bBefore = snap(rb);

    const winScore = Math.max(g.aScore, g.bScore);
    const loseScore = Math.min(g.aScore, g.bScore);
    const [[newWinner], [newLoser]] = rate([[winnerR], [loserR]], {
      ...RATE_OPTS,
      score: normScores(winScore, loseScore, g.target),
    });
    ratings.set(winnerId, newWinner);
    ratings.set(loserId, newLoser);

    const aAfter = snap(ratings.get(g.aId)!);
    const bAfter = snap(ratings.get(g.bId)!);
    processed.push({
      ...g,
      winnerId,
      loserId,
      winnerScore: winScore,
      loserScore: loseScore,
      margin: winScore - loseScore,
      isDeuce: winScore > g.target,
      winnerPreProb,
      aBefore,
      bBefore,
      aAfter,
      bAfter,
    });
    history.get(g.aId)!.push({ gameIndex: idx + 1, playedAt: g.playedAt, ...hist(aAfter) });
    history.get(g.bId)!.push({ gameIndex: idx + 1, playedAt: g.playedAt, ...hist(bAfter) });
  });

  const playerStats: Record<string, PlayerStats> = {};
  for (const p of players) {
    playerStats[p.id] = computePlayerStats(p, processed, ratings, history, players);
  }

  const ranked = Object.values(playerStats).sort(
    (a, b) =>
      Number(a.provisional) - Number(b.provisional) || b.rating.exposed - a.rating.exposed,
  );
  ranked.forEach((s, i) => {
    s.rank = i + 1;
    const h = s.ratingHistory;
    if (h.length > 5) {
      const past = h[h.length - 6].exposed;
      const now = s.rating.exposed;
      s.rankDelta = Math.abs(now - past) < 0.05 ? 0 : now > past ? 1 : -1;
    }
  });

  const matchups: Record<string, Record<string, number>> = {};
  for (const a of players) {
    matchups[a.id] = {};
    for (const b of players) {
      if (b.id === a.id) continue;
      matchups[a.id][b.id] = winProb(ratings.get(a.id)!, ratings.get(b.id)!);
    }
  }

  return {
    players,
    games,
    processedGames: processed,
    playerStats,
    rankedIds: ranked.map((s) => s.player.id),
    league: leagueStats(processed, ranked),
    matchups,
    config: { model: "BradleyTerryFull", margin: MARGIN, provisionalThreshold: PROVISIONAL_THRESHOLD },
  };
}

/** Pick the element with the maximum (key[0], key[1]) tuple, like Python's max(..., key=). */
function maxBy<T>(xs: T[], key: (x: T) => [number, number]): T | null {
  let best: T | null = null;
  let bestKey: [number, number] | null = null;
  for (const x of xs) {
    const k = key(x);
    if (!bestKey || k[0] > bestKey[0] || (k[0] === bestKey[0] && k[1] > bestKey[1])) {
      best = x;
      bestKey = k;
    }
  }
  return best;
}

function computePlayerStats(
  p: Player,
  processed: ProcessedGame[],
  ratings: Map<string, OSRating>,
  history: Map<string, PlayerStats["ratingHistory"]>,
  players: Player[],
): PlayerStats {
  const pid = p.id;
  const mine = processed.filter((g) => g.aId === pid || g.bId === pid);
  const wins = mine.filter((g) => g.winnerId === pid);
  const losses = mine.filter((g) => g.loserId === pid);

  const pointsFor = mine.reduce((acc, g) => acc + (g.aId === pid ? g.aScore : g.bScore), 0);
  const pointsAgainst = mine.reduce((acc, g) => acc + (g.aId === pid ? g.bScore : g.aScore), 0);

  let currentStreak: PlayerStats["currentStreak"] = null;
  let longestW = 0;
  let longestL = 0;
  let runKind: "W" | "L" | null = null;
  let runLen = 0;
  for (const g of mine) {
    const kind: "W" | "L" = g.winnerId === pid ? "W" : "L";
    runLen = kind === runKind ? runLen + 1 : 1;
    runKind = kind;
    if (kind === "W") longestW = Math.max(longestW, runLen);
    else longestL = Math.max(longestL, runLen);
  }
  if (runKind) currentStreak = { kind: runKind, length: runLen };

  const biggestWin = maxBy(wins, (g) => [g.margin, g.winnerScore]);
  const biggestLoss = maxBy(losses, (g) => [g.margin, g.winnerScore]);
  const deuce = mine.filter((g) => g.isDeuce);

  const h2h: HeadToHead[] = [];
  for (const opp of players) {
    const oid = opp.id;
    if (oid === pid) continue;
    const vs = mine.filter((g) => g.aId === oid || g.bId === oid);
    const w = vs.filter((g) => g.winnerId === pid);
    const l = vs.filter((g) => g.loserId === pid);
    const pf = vs.reduce((acc, g) => acc + (g.aId === pid ? g.aScore : g.bScore), 0);
    const pa = vs.reduce((acc, g) => acc + (g.aId === pid ? g.bScore : g.aScore), 0);
    h2h.push({
      opponentId: oid,
      wins: w.length,
      losses: l.length,
      pointsFor: pf,
      pointsAgainst: pa,
      winProb: winProb(ratings.get(pid)!, ratings.get(oid)!),
      avgMarginInWins: avg(w.map((g) => g.margin)),
      avgMarginInLosses: avg(l.map((g) => g.margin)),
      lastResults: vs.slice(-5).map((g) => (g.winnerId === pid ? "W" : "L")),
    });
  }

  const contested = h2h.filter((m) => m.wins + m.losses >= 2);
  const byWinPct = [...contested].sort(
    (a, b) => a.wins / (a.wins + a.losses) - b.wins / (b.wins + b.losses),
  );
  const nemesis = byWinPct.length ? byWinPct[0] : null;
  const bestMatchup = byWinPct.length ? byWinPct[byWinPct.length - 1] : null;

  const r = snap(ratings.get(pid)!);
  return {
    player: p,
    rating: r,
    rank: 0,
    provisional: mine.length < PROVISIONAL_THRESHOLD,
    games: mine.length,
    wins: wins.length,
    losses: losses.length,
    winRate: mine.length ? wins.length / mine.length : 0,
    currentStreak,
    longestWinStreak: longestW,
    longestLossStreak: longestL,
    pointsFor,
    pointsAgainst,
    pointDiff: pointsFor - pointsAgainst,
    avgPointsFor: mine.length ? pointsFor / mine.length : 0,
    avgPointsAgainst: mine.length ? pointsAgainst / mine.length : 0,
    avgMarginInWins: avg(wins.map((g) => g.margin)),
    avgMarginInLosses: avg(losses.map((g) => g.margin)),
    biggestWin,
    biggestLoss,
    deuceGames: deuce.length,
    deuceRecord: {
      wins: deuce.filter((g) => g.winnerId === pid).length,
      losses: deuce.filter((g) => g.loserId === pid).length,
    },
    recordBy11: {
      wins: wins.filter((g) => g.target === 11).length,
      losses: losses.filter((g) => g.target === 11).length,
    },
    recordBy21: {
      wins: wins.filter((g) => g.target === 21).length,
      losses: losses.filter((g) => g.target === 21).length,
    },
    upsetsPulled: wins.filter((g) => g.winnerPreProb < 0.5).length,
    upsetsSuffered: losses.filter((g) => 1 - g.winnerPreProb < 0.5).length,
    peakRating: Math.max(...history.get(pid)!.map((h) => h.exposed)),
    form: mine.slice(-10).map((g) => (g.winnerId === pid ? "W" : "L")),
    ratingHistory: history.get(pid)!,
    headToHead: h2h,
    nemesis,
    bestMatchup,
    rankDelta: null,
  };
}

function leagueStats(processed: ProcessedGame[], ranked: PlayerStats[]): LeagueStats {
  const byUpset = [...processed].sort((a, b) => a.winnerPreProb - b.winnerPreProb);
  const byMargin = [...processed].sort(
    (a, b) => a.margin - b.margin || b.winnerScore - a.winnerScore,
  );
  const byBlowout = [...processed].sort((a, b) => b.margin - a.margin);

  let longestActive: LeagueStats["longestActiveStreak"] = null;
  for (const s of ranked) {
    const cs = s.currentStreak;
    if (cs && cs.kind === "W") {
      if (longestActive === null || cs.length > longestActive.length) {
        longestActive = { player: s.player, length: cs.length };
      }
    }
  }

  let mostActive: PlayerStats | null = null;
  for (const s of ranked) {
    if (mostActive === null || s.games > mostActive.games) mostActive = s;
  }

  return {
    totalGames: processed.length,
    totalPoints: processed.reduce((acc, g) => acc + g.aScore + g.bScore, 0),
    deuceGames: processed.filter((g) => g.isDeuce).length,
    gamesTo11: processed.filter((g) => g.target === 11).length,
    gamesTo21: processed.filter((g) => g.target === 21).length,
    avgMargin: processed.length
      ? processed.reduce((acc, g) => acc + g.margin, 0) / processed.length
      : 0,
    biggestUpsets: byUpset.filter((g) => g.winnerPreProb < 0.5).slice(0, 5),
    closestGames: byMargin.slice(0, 5),
    mostLopsided: byBlowout.slice(0, 5),
    longestActiveStreak: longestActive,
    mostActive: mostActive ? { player: mostActive.player, games: mostActive.games } : null,
  };
}

File: backend/engine.py
"""Rating engine: replays the game log chronologically through openskill.py.

Model: Bradley-Terry (full pairing) with margin-of-victory enabled.
Scores are normalized to an 11-point scale before rating so that a 21-15
game and an 11-8 game carry comparable margin information.
"""

from __future__ import annotations

import math
import os
from typing import Any

from openskill.models import BradleyTerryFull

PROVISIONAL_THRESHOLD = 5
MARGIN = float(os.environ.get("MARGIN", "2.0"))

_model = BradleyTerryFull(margin=MARGIN)


def _snap(r) -> dict[str, float]:
    return {"mu": r.mu, "sigma": r.sigma, "exposed": r.mu - 3 * r.sigma}


def _norm_scores(a_score: int, b_score: int, target: int) -> list[float]:
    """Scale scores to an 11-point-equivalent so margins are comparable
    across game lengths."""
    k = 11.0 / target
    return [a_score * k, b_score * k]


def compute(players: list[dict], games: list[dict]) -> dict[str, Any]:
    ratings = {p["id"]: _model.rating(name=p["id"]) for p in players}
    history: dict[str, list[dict]] = {
        p["id"]: [
            {"gameIndex": 0, "playedAt": p["createdAt"], **_hist(_snap(ratings[p["id"]]))}
        ]
        for p in players
    }

    processed: list[dict] = []
    for idx, g in enumerate(sorted(games, key=lambda x: x["playedAt"])):
        ra, rb = ratings.get(g["aId"]), ratings.get(g["bId"])
        if ra is None or rb is None:
            continue

        a_won = g["aScore"] > g["bScore"]
        winner_id = g["aId"] if a_won else g["bId"]
        loser_id = g["bId"] if a_won else g["aId"]
        winner_r = ra if a_won else rb
        loser_r = rb if a_won else ra

        winner_pre_prob = _model.predict_win([[winner_r], [loser_r]])[0]
        a_before, b_before = _snap(ra), _snap(rb)

        win_score = max(g["aScore"], g["bScore"])
        lose_score = min(g["aScore"], g["bScore"])
        [[new_winner], [new_loser]] = _model.rate(
            [[winner_r], [loser_r]],
            scores=_norm_scores(win_score, lose_score, g["target"]),
        )
        ratings[winner_id], ratings[loser_id] = new_winner, new_loser

        a_after, b_after = _snap(ratings[g["aId"]]), _snap(ratings[g["bId"]])
        processed.append(
            {
                **g,
                "winnerId": winner_id,
                "loserId": loser_id,
                "winnerScore": win_score,
                "loserScore": lose_score,
                "margin": win_score - lose_score,
                "isDeuce": win_score > g["target"],
                "winnerPreProb": winner_pre_prob,
                "aBefore": a_before,
                "bBefore": b_before,
                "aAfter": a_after,
                "bAfter": b_after,
            }
        )
        history[g["aId"]].append(
            {"gameIndex": idx + 1, "playedAt": g["playedAt"], **_hist(a_after)}
        )
        history[g["bId"]].append(
            {"gameIndex": idx + 1, "playedAt": g["playedAt"], **_hist(b_after)}
        )

    player_stats = {
        p["id"]: _player_stats(p, processed, ratings, history, players)
        for p in players
    }

    ranked = sorted(
        player_stats.values(),
        key=lambda s: (s["provisional"], -s["rating"]["exposed"]),
    )
    for i, s in enumerate(ranked):
        s["rank"] = i + 1
        hist = s["ratingHistory"]
        if len(hist) > 5:
            past, now = hist[-6]["exposed"], s["rating"]["exposed"]
            s["rankDelta"] = 0 if abs(now - past) < 0.05 else (1 if now > past else -1)

    matchups = {
        a["id"]: {
            b["id"]: _model.predict_win([[ratings[a["id"]]], [ratings[b["id"]]]])[0]
            for b in players
            if b["id"] != a["id"]
        }
        for a in players
    }

    return {
        "players": players,
        "games": games,
        "processedGames": processed,
        "playerStats": player_stats,
        "rankedIds": [s["player"]["id"] for s in ranked],
        "league": _league_stats(processed, ranked),
        "matchups": matchups,
        "config": {"model": "BradleyTerryFull", "margin": MARGIN, "provisionalThreshold": PROVISIONAL_THRESHOLD},
    }


def _hist(snap: dict) -> dict:
    return {"exposed": snap["exposed"], "mu": snap["mu"]}


def _avg(xs: list[float]) -> float | None:
    return sum(xs) / len(xs) if xs else None


def _player_stats(p, processed, ratings, history, players) -> dict:
    pid = p["id"]
    mine = [g for g in processed if pid in (g["aId"], g["bId"])]
    wins = [g for g in mine if g["winnerId"] == pid]
    losses = [g for g in mine if g["loserId"] == pid]

    points_for = sum(g["aScore"] if g["aId"] == pid else g["bScore"] for g in mine)
    points_against = sum(g["bScore"] if g["aId"] == pid else g["aScore"] for g in mine)

    current_streak = None
    longest_w = longest_l = 0
    run_kind, run_len = None, 0
    for g in mine:
        kind = "W" if g["winnerId"] == pid else "L"
        run_len = run_len + 1 if kind == run_kind else 1
        run_kind = kind
        if kind == "W":
            longest_w = max(longest_w, run_len)
        else:
            longest_l = max(longest_l, run_len)
    if run_kind:
        current_streak = {"kind": run_kind, "length": run_len}

    biggest_win = max(wins, key=lambda g: (g["margin"], g["winnerScore"]), default=None)
    biggest_loss = max(losses, key=lambda g: (g["margin"], g["winnerScore"]), default=None)
    deuce = [g for g in mine if g["isDeuce"]]

    h2h = []
    for opp in players:
        oid = opp["id"]
        if oid == pid:
            continue
        vs = [g for g in mine if oid in (g["aId"], g["bId"])]
        w = [g for g in vs if g["winnerId"] == pid]
        l = [g for g in vs if g["loserId"] == pid]
        pf = sum(g["aScore"] if g["aId"] == pid else g["bScore"] for g in vs)
        pa = sum(g["bScore"] if g["aId"] == pid else g["aScore"] for g in vs)
        h2h.append(
            {
                "opponentId": oid,
                "wins": len(w),
                "losses": len(l),
                "pointsFor": pf,
                "pointsAgainst": pa,
                "winProb": _model.predict_win([[ratings[pid]], [ratings[oid]]])[0],
                "avgMarginInWins": _avg([g["margin"] for g in w]),
                "avgMarginInLosses": _avg([g["margin"] for g in l]),
                "lastResults": ["W" if g["winnerId"] == pid else "L" for g in vs[-5:]],
            }
        )

    contested = [m for m in h2h if m["wins"] + m["losses"] >= 2]
    by_win_pct = sorted(contested, key=lambda m: m["wins"] / (m["wins"] + m["losses"]))
    nemesis = by_win_pct[0] if by_win_pct else None
    best_matchup = by_win_pct[-1] if by_win_pct else None

    snap = _snap(ratings[pid])
    return {
        "player": p,
        "rating": snap,
        "rank": 0,
        "provisional": len(mine) < PROVISIONAL_THRESHOLD,
        "games": len(mine),
        "wins": len(wins),
        "losses": len(losses),
        "winRate": len(wins) / len(mine) if mine else 0.0,
        "currentStreak": current_streak,
        "longestWinStreak": longest_w,
        "longestLossStreak": longest_l,
        "pointsFor": points_for,
        "pointsAgainst": points_against,
        "pointDiff": points_for - points_against,
        "avgPointsFor": points_for / len(mine) if mine else 0.0,
        "avgPointsAgainst": points_against / len(mine) if mine else 0.0,
        "avgMarginInWins": _avg([g["margin"] for g in wins]),
        "avgMarginInLosses": _avg([g["margin"] for g in losses]),
        "biggestWin": biggest_win,
        "biggestLoss": biggest_loss,
        "deuceGames": len(deuce),
        "deuceRecord": {
            "wins": sum(1 for g in deuce if g["winnerId"] == pid),
            "losses": sum(1 for g in deuce if g["loserId"] == pid),
        },
        "recordBy11": {
            "wins": sum(1 for g in wins if g["target"] == 11),
            "losses": sum(1 for g in losses if g["target"] == 11),
        },
        "recordBy21": {
            "wins": sum(1 for g in wins if g["target"] == 21),
            "losses": sum(1 for g in losses if g["target"] == 21),
        },
        "upsetsPulled": sum(1 for g in wins if g["winnerPreProb"] < 0.5),
        "upsetsSuffered": sum(1 for g in losses if 1 - g["winnerPreProb"] < 0.5),
        "peakRating": max(h["exposed"] for h in history[pid]),
        "form": ["W" if g["winnerId"] == pid else "L" for g in mine[-10:]],
        "ratingHistory": history[pid],
        "headToHead": h2h,
        "nemesis": nemesis,
        "bestMatchup": best_matchup,
        "rankDelta": None,
    }


def _league_stats(processed: list[dict], ranked: list[dict]) -> dict:
    by_upset = sorted(processed, key=lambda g: g["winnerPreProb"])
    by_margin = sorted(processed, key=lambda g: (g["margin"], -g["winnerScore"]))
    by_blowout = sorted(processed, key=lambda g: -g["margin"])

    longest_active = None
    for s in ranked:
        cs = s["currentStreak"]
        if cs and cs["kind"] == "W":
            if longest_active is None or cs["length"] > longest_active["length"]:
                longest_active = {"player": s["player"], "length": cs["length"]}

    most_active = max(ranked, key=lambda s: s["games"], default=None)
    return {
        "totalGames": len(processed),
        "totalPoints": sum(g["aScore"] + g["bScore"] for g in processed),
        "deuceGames": sum(1 for g in processed if g["isDeuce"]),
        "gamesTo11": sum(1 for g in processed if g["target"] == 11),
        "gamesTo21": sum(1 for g in processed if g["target"] == 21),
        "avgMargin": (sum(g["margin"] for g in processed) / len(processed)) if processed else 0.0,
        "biggestUpsets": [g for g in by_upset if g["winnerPreProb"] < 0.5][:5],
        "closestGames": by_margin[:5],
        "mostLopsided": by_blowout[:5],
        "longestActiveStreak": longest_active,
        "mostActive": {"player": most_active["player"], "games": most_active["games"]} if most_active else None,
    }
